components/eamxx/src/dynamics/homme/eamxx_homme_iop.cpp

```diff
@@ -0,0 +1,408 @@
+
+    const auto weight = (pint(k) - pmid_km1)/(pmid_k - pmid_km1);
+    omega_int(k).set(range_pack>=1 and range_pack<=nlevs-1,
+                      weight*omega_k + (1-weight)*omega_km1);
+  });
+  omega_int(0)[0] = 0;
+  omega_int(nlevs/Pack::n)[nlevs%Pack::n] = 0;
+
+  // Compute delta views for u, v, T, and Q (e.g., u(k+1) - u(k), k=0,...,nlevs-2)
+  ColOps::compute_midpoint_delta(team, nlevs-1, u, delta_u);
+  ColOps::compute_midpoint_delta(team, nlevs-1, v, delta_v);
+  ColOps::compute_midpoint_delta(team, nlevs-1, T, delta_T);
+  for (int iq=0; iq<n_q_tracers; ++iq) {
+    auto tracer       = Kokkos::subview(Q,       iq, Kokkos::ALL());
+    auto delta_tracer = Kokkos::subview(delta_Q, iq, Kokkos::ALL());
+    ColOps::compute_midpoint_delta(team, nlevs-1, tracer, delta_tracer);
+  }
+  team.team_barrier();
+
+  // Compute updated temperature, horizontal winds, and tracers
+  Kokkos::parallel_for(Kokkos::TeamVectorRange(team, nlev_packs), [&] (const int k) {
+    auto range_pack = ekat::range<IntPack>(k*Pack::n);
+
+    // Get delta(k-1) packs. We need a range pack
+    // that does not contain 0 so that we do not
+    // attempt to access k=-1 index.
+    auto range_pack_m1 = range_pack;
+    range_pack_m1.set(range_pack_m1<1, 1);
+    Pack delta_u_k, delta_u_km1,
+          delta_v_k, delta_v_km1,
+          delta_T_k, delta_T_km1;
+    ekat::index_and_shift<-1>(s_delta_u, range_pack_m1, delta_u_k, delta_u_km1);
+    ekat::index_and_shift<-1>(s_delta_v, range_pack_m1, delta_v_k, delta_v_km1);
+    ekat::index_and_shift<-1>(s_delta_T, range_pack_m1, delta_T_k, delta_T_km1);
+
+    // Get omega_int(k+1) pack. We don't need a specialized
+    // range pack since omega_int contains nlevs+1 entries.
+    Pack omega_int_k, omega_int_kp1;
+    ekat::index_and_shift<1>(s_omega_int, range_pack, omega_int_k, omega_int_kp1);
+
+    auto at_top = range_pack==0;
+    auto at_bot = range_pack==nlevs-1;
+
+    auto fac = dt/(2*pdel(k));
+
+    // Update u
+    u(k).set(at_top,                    u(k) - fac*omega_int_kp1*delta_u_k);
+    u(k).set(at_bot,                    u(k) - fac*omega_int_k*delta_u_km1);
+    u(k).set(not at_top and not at_bot, u(k) - fac*(omega_int_kp1*delta_u_k + omega_int_k*delta_u_km1));
+
+    // Update v
+    v(k).set(at_top,                    v(k) - fac*omega_int_kp1*delta_v_k);
+    v(k).set(at_bot,                    v(k) - fac*omega_int_k*delta_v_km1);
+    v(k).set(not at_top and not at_bot, v(k) - fac*(omega_int_kp1*delta_v_k + omega_int_k*delta_v_km1));
+
+    // Update T
+    const auto T_k = T(k);
+    T(k).set(at_top,                    T_k - fac*omega_int_kp1*delta_T_k);
+    T(k).set(at_bot,                    T_k - fac*omega_int_k*delta_T_km1);
+    T(k).set(not at_top and not at_bot, T_k - fac*(omega_int_kp1*delta_T_k + omega_int_k*delta_T_km1));
+    // Add thermal expansion term due to LS vertical advection
+    T(k) += dt*omega(k)*T_k*Rair/(Cpair*pmid(k));;
```

Comment:
Again, it's best to do all the scalar ops at once, since they are cheaper than vector ops. Any chance we can turn this into this?
```suggestion
  T(k) += (dt*Rair/Cpair) * omega(k)*T_k/pmid(k) ;
```
If you need to keep BFB with F90, for a <5 lines snippet of code, it's ok to do something like
```c++
#ifdef EAMXX_BFB_TESTING
  // impl with same order of ops as in f90
#else
  // impl with better order of ops
#endif
```
Note: I don't recall if we have that macro, but we should have something equivalent.

Reply:
Also, in the update of `T(k)`, is it intended to use `T_k` rather than `T(k)` on the rhs?

Reply:
Finally, from the Tk update formula, putting the pieces together, it seems we do
```
Tk = Tk - fac*blah + dt*Tk*blah2
```
Maybe we could then do
```
auto& Tk = T(k);
Tk *= (1+dt*blah2);
ekat_masked_loop (at_top,s)
  Tk[s] -= fac*blah_top;
ekat_masked_loop (at_bot,s)
  Tk[s] -= fac*blah_bot;
ekat_masked_loop (at_mid,s)
  Tk[s] -= fac*blah_mid;
```

Reply:
If we impl-ed an `update` method for `Pack` similar to `Pack::set`, we could do
```
auto& Tk = T(k);
Tk *= (1+dt*blah2);
Tk.update<Subtract>(not at_top,fac*blah_not_top);
Tk.udpate<Subtract>(not at_bot,fac*blah_not_bot);
```

Reply:
What would be the point of going with the `ekat_masked_loop` instead of using `Pack::set(mask,)`?

And yes, an update method would be great. I think there are a few places that could already use it.

Reply:
To answer an earlier question, the original `T(k)` needed to be used in
```
T(k) += dt*omega(k)*T_k*Rair/(Cpair*pmid(k));
```
but now I followed your suggestion and just first scaled by `(1+dt*blah2)`.